src/clj/clj_diff/miller.clj

```diff
@@ -41,31 +49,41 @@
   in the map fp. Returns an updated fp map for p. a and b are the two
   sequences and n and m are their lengths respectively. delta is the
   diagonal of the sink and is equal to n - m."
-  [a b n m delta p fp]
-  (reduce (fn [fp next-k]
-            (assoc fp next-k (snake a b n m next-k fp)))
-          fp
-          (p-band-diagonals p delta)))
+  [a b n m delta p fp & [opts]]
+  (let [opts (or opts {})]
+    (persistent! (reduce (fn [fp next-k]
+                           (assoc! fp next-k (snake a b n m next-k fp opts)))
+                   (transient fp)
+                   (p-band-diagonals p delta)))))
 
 (defn ses
   "Find the size of the shortest edit script (ses). Returns a 3-tuple of the
   size of the ses, the delta value (which is the diagonal of the sink)
   and the fp map. The optimal path from source to sink can be constructed from
   this information."
-  [a b]
+  [a b & [opts]]
   {:pre [(>= (count a) (count b))]}
-  (let [n (dec (count a))
-        m (dec (count b))
-        delta (- n m)]
-    (loop [p 0
+  (let [opts (or opts {})
+        limit-p (:limit-p opts)
+        n       (dec (count a))
+        m       (dec (count b))
+        delta   (- n m)]
+    (loop [p  0
            fp {}]
-      (if (= (-> (get fp (dec p) {})
-                 (get delta))
-             n)
-        [(dec p) delta fp]
-        (recur (inc p)
-               (assoc fp p
-                      (search-p-band a b n m delta p (get fp (dec p) {}))))))))
+
+      (cond
+        (and limit-p
+             (> p limit-p))
+        #_=> [-1 delta fp]
+
+        (= (-> (fast/val-at fp (dec p) {})
+               (fast/val-at delta))
+           n)
+        #_=> [(dec p) delta fp]
```

Comment:
More commented out code

Reply:
Apparently I can't read properly and these are fine

Reply:
cheers